### DISCUSSION TITLE
Can I define a closure on the AppState object for usage in a globalTimeout-Event?

### DISCUSSION THREAD

**Discussion by PhilippMDoerner** [can-lehmann/owlkettle#150] [Q&A] [answered]
Heyho, for Appster's message integration I would like to be able to use "signalling"-type syntax, as in I want users to define callbacks that then get called when a message arrives. The callback gets executed within this proc:
```nim
proc addServerListener*[OwlkettleApp: Viewable, SMsg, CMsg](
  app: OwlkettleApp, 
  data: ServerData[SMsg, CMsg]
) =  
  ## Adds a callback function to the GTK app that checks every 5 ms whether the 
  ## server sent a new message. Triggers a UI update if that is the case.
  proc listener(): bool =
    let msg = data.hub.readServerMsg()
    if msg.isSome():
      app.msg = msg.get()
      discard app.redraw()
    
    const KEEP_LISTENER_ACTIVE = true
    return KEEP_LISTENER_ACTIVE

  discard addGlobalTimeout(data.sleepMs, listener)
```
Basically I want to enable the user to define a proc that gets called when the backend-thread sends a message to the frontend-thread.
I noticed that this is already somewhat possible because this syntax here compiles:
```nim

viewable App:
  server: ExampleServer
  msg: ServerMessage
  inputText: string
  receivedMessages: seq[string]
  
  proc messageHandler(msg: ServerMessage) = echo "Received a message!"
```
to
```nim
AppState = ref object of Viewable
    server*: ExampleServer
    msg*: ServerMessage
    inputText*: string
    receivedMessages*: seq[string]
    messageHandler*: Event[proc (msg: ServerMessage)]
```

The next question is - how do I get an actual proc into that field, because this just defines the field, it doesn't fill it, it is always nil:
```nim
method view(app: AppState): Widget =
  echo app.messageHandler.isNil() # true

  result = gui:
    Window:
      defaultSize = (500, 150)
      title = "Client Server Example"
``` 

Or do I have any alternatives to store a callback in an AppState?
Because mere assignment doesn't compile:
```nim
viewable App:
  server: ExampleServer
  msg: ServerMessage
  inputText: string
  receivedMessages: seq[string]
  handler: proc(state: AppState, msg: ServerMessage)
  ```
leads to: `Error: nestable statement requires indentation` (on the line with `handler`, before any codegen on owlkettle's part can occur).

Edit: Thinking about it, the answer to this likely should lead to a chapter in the docs on how to have custom Owlkettle events in a viewable.

**Comment by PhilippMDoerner** [comment]
I managed to bypass the reason I wanted that kind of callback for by just making use of the fact that the proc I need it in is a generic and thus I can just write the proc as if it existed in there and use `mixin <thatprocname>` and it'll look for the proc in the scope of the callsite - which means I don't need a closure in the type, I just need the proc defined at the call-site.

Which I can do, I'll just need to add some more codegen and clean up the appster API a bit, which works fine.

**Comment by can-lehmann** [comment, marked as answer]
>  I noticed that this is already somewhat possible because this syntax here compiles:

The procedure body is simply ignored here. It will now throw an error in this case https://github.com/can-lehmann/owlkettle/commit/5559d63ef6f7207dcd6325a6bc9c17e58d7ca15a.

**Reply by PhilippMDoerner** [reply]
I love me some validation :smile: 

Thanks for adding it! I highly value macros validating what users provide and giving useful error messages when using them in invalid ways, simply because it cuts down on confusion *so much*.
